File: src/com/awiselow/redup/DimTileService.java
package com.awiselow.redup;

import android.content.Intent;
import android.os.Build;
import android.service.quicksettings.QSTileService;
import android.service.quicksettings.Tile;

public class DimTileService extends QSTileService {
    private boolean active = false;

    @Override
    public void onTileAdded() { updateTile(); }
    @Override
    public void onStartListening() { updateTile(); }

    @Override
    public void onClick() {
        active = !active;
        Intent srv = new Intent(this, DimService.class);
        if (active) {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O)
                startForegroundService(srv);
            else startService(srv);
        } else stopService(srv);
        updateTile();
    }

    private void updateTile() {
        Tile t = getQsTile();
        if (t != null) {
            t.setState(active ? Tile.STATE_ACTIVE : Tile.STATE_INACTIVE);
            t.updateTile();
        }
    }
}

File: app/src/main/java/com/awiselow/redup/DimTileService.kt
package com.awiselow.redup

import android.content.Intent
import android.service.quicksettings.Tile
import android.service.quicksettings.TileService

class DimTileService : TileService() {

    override fun onTileAdded() {
        updateTile(isRunning())
    }

    override fun onStartListening() {
        updateTile(isRunning())
    }

    override fun onClick() {
        if (isRunning()) {
            sendBroadcast(Intent("com.awiselow.redup.EXIT"))
            updateTile(false)
        } else {
            val intent = Intent(this, DimService::class.java).apply {
                putExtra("alpha", 80)
            }
            startForegroundService(intent)
            updateTile(true)
        }
    }

    private fun updateTile(active: Boolean) {
        qsTile?.apply {
            state = if (active) Tile.STATE_ACTIVE else Tile.STATE_INACTIVE
            label = "Redup"
            updateTile()
        }
    }

    private fun isRunning(): Boolean {
        return getSharedPreferences("redup_prefs", MODE_PRIVATE)
            .getBoolean("is_active", false)
    }
}
